### ISSUE TITLE
"Mods not having a mod.conf file with the name is deprecated." breaks compatibility

### ISSUE THREAD

**Issue by nephele-gh** [minetest/minetest#13166]
I would like to note that having to have a /specific/ name in mod.conf breaks the usecase of simply cloning mods to compare behaviour between them.

It's not that hard to make mods work with arbitrary modnames, and i have done that with all my gamemode specific mods.
If mod.conf is the only way there should be an option to inherit the modname from the folder as before without having to specify it explicitly.

**Comment by fluxionary** [comment]
> I would like to note that having to have a /specific/ name in mod.conf breaks the usecase of simply cloning mods to compare behaviour between them.
>
> It's not that hard to make mods work with arbitrary modnames, and i have done that with all my gamemode specific mods. If mod.conf is the only way there should be an option to inherit the modname from the folder as before without having to specify it explicitly.

do you mean, having two or more mods w/ the same name defined in mod.conf, and using them all simultaneously? i try to account for my mods being renamed to some extent, but this issue makes me aware that i'm not doing it properly in a lot of places. thanks for the heads-up.

at the same time, comparing forks/implementations of the same mod by using them all simultaneously seems like a recipe for incredible trouble for lots of reasons, particularly if anything is overriding any of the builtin API or other API mods. i'd recommend doing your comparisons in a different way.

probably most importantly, this isn't an engine issue - and so not something that can be fixed here - but related to "best practices" for writing mods. rubenwardy has written The Book on mods for the moment. protocol on how to maintain mods amongst the rest of the community is a topic that i want to contribute to at some point, but i'm not quite ready to write that chapter yet.

**Comment by nephele-gh** [comment]
> trouble for lots of reasons, particularly if anything is overriding any of the builtin API or other API mods. i'd recommend doing your comparisons in a different way.

This is not the case, and I dislike bending over backwards because of some arbitrary decision that it is now the wrong way.

> protocol on how to maintain mods amongst the rest of the community

I am not interested in that, I develop my own gamemode.

That I am expected to migrate all my gamemodes mods now under the threat of broken compatibility later (also with depends.txt) is particularily annoying since the format isn't even documented. It's just a list of key.

**Comment by rollerozxa** [comment]
The mod.conf format *is* in fact documented: https://minetest.gitlab.io/minetest/mods/#modconf

I don't see any case that someone would want to... change the name of a mod, just like that? Unless it is accidental, if someone downloads the zip from GitHub and extracts it into a folder named cool_mod-master, and then mod.conf is useful for the engine to detect the proper name.

**Comment by nephele-gh** [comment]
> The mod.conf format is in fact documented: https://minetest.gitlab.io/minetest/mods/#modconf

This is the same documentation as in lua_api.txt, and it does not document the format at all. It only provided a list of keys... which would be usefull if the actual syntax was documented.

**Comment by rollerozxa** [comment]
Wrong, that is in fact lua_api.txt generated into HTML format, and the Settings key-value syntax is also documented: https://minetest.gitlab.io/minetest/class-reference/#format

**Comment by rollerozxa** [comment]
Misread that as "This is not the same documentation as in lua_api.txt". The mod.conf format is still documented though, it uses the same syntax as all other files that use the Settings format (minetest.conf, world.mt, any other config files mods use...) and the documentation mentions this.

**Comment by rollerozxa** [comment]
At the top of the mod.conf section:

> **mod.conf**
> A Settings file that provides meta information about the mod.

**Comment by appgurueu** [comment]
First of all, I'm very certain that no compatibility is broken (yet), since this is just a *deprecation warning*; Minetest should still use the folder name if you don't specify a name in `mod.conf`

> I would like to note that having to have a /specific/ name in mod.conf breaks the usecase of simply cloning mods to compare behaviour between them.

1. As has been pointed out already, this is a very error-prone strategy that relies on
   1. The mod being compatible with itself in a different version
   2. Other mods being able to deal with multiple versions of... the same mod? active at the same time
   3. The mod being written in such a way that it can deal with an arbitrary modname (using `minetest.get_current_modname()`, which I agree may be best practice, but in )
2. You can still trivially "fix" your use case by adding a `echo "name = blarg" >> mod.conf` step to your workflow.
3. I would suggest you switch to testing the different versions in subsequent runs rather than in the same run. This may take slightly more time (mostly just + startup/shutdown time, considering we humans aren't really able to do two things simultaneously). If you wanted to be really fancy, you could have two different Minetest instances running side by side, using two different mod folders, which would contain two different versions of your mods (but still with the same name). That way you would still get side by side comparisons and assuming you have 2+ CPU cores you likely won't lose any time ;)

> It's not that hard to make mods work with arbitrary modnames, and i have done that with all my gamemode specific mods.

1. The current "average case" is mods hardcoding modnames. I agree that this mostly is poor practice, but it means that usually, the job of the engine should be to make this average case *work*. And for that the engine has to enforce modnames which must be specified in `mod.conf`. The "average error" back in the 0.4.x days used to be people downloading mod archives with `-master` in their name, which is what was properly addressed using `mod.conf`.
2. Sure, you can parameterize all your "names" in terms of the modname. That is indeed easy (although still not widespread unfortunately since modders are lazy and don't consider the possibility of having to change the name later on). When you consider APIs however it stops being this easy: An API is required to have a certain name for other mods to be able to work with it; otherwise these other mods couldn't even identify the presence of the API. This (1) completely breaks your "same mod, different versions" testing strategy, because only one mod can set this global API table; which API table are other mods supposed to use? (2) `depends` and `optional_depends` means names *must* be fixed for these to work.

In conclusion, the only type of mods this would make sense for are "leaf/content mods": Mods which no other mods depend on and which do not provide an API (e.g. mods just registering nodes) to be compared side by side using an error-prone testing strategy. This makes this "use case" rather narrow and convoluted (IMO).

> If mod.conf is the only way there should be an option to inherit the modname from the folder as before without having to specify it explicitly.

Debatable. Either way this makes this issue a *feature request* rather than a *bug report*.

**TL;DR:** No compatibility breakage, not a bug, merely a (IMO low priority) *feature request*.

**Comment by appgurueu** [comment]
> Point 2 is wrong. If I am forced to have a modname in the conf then appending one would not work. Maybe you are looking for sed? Anyway, I'm not willing to add a preprocessing stage to "build" my gamemode because of an api change.

Incorrect. Later entries in `Settings` files will override previous ones.


> Point 3 is only partially correct, discovery of which api names to use at runtime is perfectly possible.

Incorrect. It is only possible using other API names, which will in turn need to be fixed.

> In any case, the solution to this( or one possible one) would be to have a flag in mod.conf to inherit the modname, it would not be particularily hard to implement and would likely have taken less time than trying to debate weather my testing strategy is "bad".

Then why didn't you open a PR?

**Comment by rubenwardy** [comment]
There's no compatibility break here as it's not a supported use case

A better argument for not having this warning is that it's noisey and has little benefit given that inexperienced users will be installing via ContentDB anyway. It'll probably lead to deprecation fatigue, where modders are more likely to just ignore deprecations
